### PR TITLE
Proper handling of MSBuild projects which don't implement the target protocol

### PR DESCRIPTION
**Pull request by smera**
This PR implements the following behavior regarding projects not implementing the target protocol:
* Leaf projects are always allowed. Only a verbose message is logged. Rationale: a leaf project doesn't need a target protocol.
* Non-leaf projects are blocked by default. A configuration option is added to opt-in to allow projects to not implement the protocol.
* If the user opted in to allow projects to not implement the protocol, a warning is issued for each project found in this scenario. Default targets are appended to the statically predicted list of targets of the reference as a heuristic to guess what targets the dependency needs. A warning is also issued for each project where non-empty default targets are appended due to this.

### CODE REVIEW

File: Public/Src/FrontEnd/MsBuild/Tracing/Log.cs [hunk 1]
```diff
@@ -226,20 +226,30 @@ internal Logger()
...
             EventTask = (ushort)Events.Tasks.Engine,
             EventOpcode = (byte)Events.Tasks.Parser,
             Keywords = (int)(Events.Keywords.UserMessage | Events.Keywords.Diagnostics))]
         public abstract void ProjectIsNotSpecifyingTheProjectReferenceProtocol(LoggingContext context, Location location, string projectName);
 
+        [GeneratedEvent(
+            (ushort)LogEventId.ProjectPredictedTargetsAlsoContainDefaultTargets,
+            EventGenerators = EventGenerators.LocalOnly,
+            EventLevel = Level.Warning,
+            Message = Events.LabeledProvenancePrefix + "Default targets were appended to the predicted target of project '{projectName}'. " +
```
Comment (erikma) [inline Public/Src/FrontEnd/MsBuild/Tracing/Log.cs]: >Default targets [](start = 56, length = 15)

Should list the default targets added in this message (DRI reduction)

File: Public/Src/FrontEnd/MsBuild.Serialization/MSBuildGraphBuilderArguments.cs [hunk 2]
```diff
@@ -56,6 +56,14 @@ public sealed class MSBuildGraphBuilderArguments
         /// </remarks>
         public IReadOnlyCollection<string> EntryPointTargets { get; }
 
+        /// <summary>
+        /// Whether projects are allows to not implement the target protocol
```
Comment (erikma) [inline Public/Src/FrontEnd/MsBuild.Serialization/MSBuildGraphBuilderArguments.cs]: >allows [](start = 33, length = 6)

allowed

File: Public/Src/FrontEnd/MsBuild.Serialization/MSBuildGraphBuilderArguments.cs [hunk 3]
```diff
@@ -56,6 +56,14 @@ public sealed class MSBuildGraphBuilderArguments
         /// </remarks>
         public IReadOnlyCollection<string> EntryPointTargets { get; }
 
+        /// <summary>
+        /// Whether projects are allows to not implement the target protocol
+        /// </summary>
+        /// <remarks>
+        /// If true, default targets are used as a heuristics
```
Comment (erikma) [inline Public/Src/FrontEnd/MsBuild.Serialization/MSBuildGraphBuilderArguments.cs]: >a heuristics [](start = 49, length = 12)

a heuristics -> heuristics

File: Public/Src/FrontEnd/MsBuild/PipConstructor.cs [hunk 4]
```diff
@@ -570,40 +600,14 @@ private bool TryAddMsBuildArguments(ProjectWithPredictions project, PipDataBuild
...
-            {
-                var targets = project.PredictedTargetsToExecute.Targets;
-                Contract.Assert(targets.Count > 0);
-                foreach (string target in targets)
-                {
-                    pipDataBuilder.Add(PipDataAtom.FromString($"/t:{target}"));
-                }
-            }
-            else
+            var targets = project.PredictedTargetsToExecute.Targets;
```
Comment (cdmihai) [inline Public/Src/FrontEnd/MsBuild/PipConstructor.cs]: Where are the default targets being added if any of the referencing projects do not specify the protocol? Can't find it.

Comment (cdmihai) [reply Public/Src/FrontEnd/MsBuild/PipConstructor.cs]: Ah, it's happening here: https://github.com/microsoft/BuildXL/pull/293/files#diff-033a479d4db36673974a933ddc616049R453

File: Public/Src/FrontEnd/MsBuild/Tracing/Log.cs (hunk 1 above)
Comment (smera) [reply Public/Src/FrontEnd/MsBuild/Tracing/Log.cs]: Yeap, good point. Added.

---
In reply to: [283011368](https://github.com/microsoft/BuildXL/pull/293#discussion_r283011368) [](ancestors = 283011368)

**Pull request merged**